apps/condo/domains/property/components/panels/Builder/MapConstructor.ts

```diff
@@ -595,10 +492,105 @@ class MapEdit extends MapView {
+        let started = false
+        let next = Number(label) + 1
+        this.map.sections.forEach(section => {
+            section.floors.slice().reverse().forEach(floor => {
+                floor.units.forEach(unit => {
+                    if (started) {
+                        if (unit.name) {
+                            if (!isNaN(Number(unit.name))) {
+                                next = Number(unit.name) + 1
+                            }
+                        } else {
+                            if (!isNaN(Number(unit.label))) {
+                                unit.label = String(next++)
+                            }
+                        }
+                    }
+                    if (unit.id === id) {
+                        started = true
+                    }
+                })
+            })
+        })
+    }
+
+    private selectedSection: BuildingSection
+
+    private selectedUnit: BuildingUnit
+
+    private generateSection (section: Partial<BuildingSectionArg>): BuildingSection {
+        let unitNumber = this.nextUnitNumber
+        const { name, minFloor, maxFloor, unitsOnFloor } = section
+        const newSection = {
+            id: String(++this.autoincrement),
+            floors: [],
+            name,
+            index: this.sections.length + 1,
+            type: null,
+        }
+        newSection.type = BuildingMapEntityType.Section
+        for (let floor = minFloor; floor <= maxFloor; floor++) {
+            if (floor === 0) {
+                continue
+            }
+            const units = []
+            for (let unitOnFloor = 0; unitOnFloor < unitsOnFloor; unitOnFloor++) {
+                let label = ' '
+                if (floor > 0) {
+                    label = String(unitNumber)
+                    unitNumber++
+                }
+                units.push({
+                    id: String(++this.autoincrement),
+                    label,
+                    type: 'unit',
+                })
+            }
+            newSection.floors.unshift({
+                id: String(++this.autoincrement),
+                index: floor,
+                name: String(floor),
+                type: BuildingMapEntityType.Floor,
+                units,
+            })
+        }
+        return newSection
+    }
+
+    private getNextUnit (id: string): BuildingUnit {
+        const units = this.map.sections.map(section => section.floors.slice(0).reverse().map(floor => floor.units)).flat(2)
+        const unitIndex = units.findIndex(unit => unit.id === id)
+        const nextIndex = unitIndex + 1
+        return units[nextIndex] || null
+    }
+
+
+    private removeFloor (sectionIdx: number, floorIndex: number): void {
+        if (!get(this.map, `sections[${sectionIdx}].floors[${floorIndex}]`, false)){
+            return
+        }
+        const floorToRemove = this.map.sections[sectionIdx].floors[floorIndex]
+        this.map.sections[sectionIdx].floors.splice(floorIndex, 1)
```

Comment:
`splice` mutates source array, are you sure you want to mutate it here ?

Reply:
@akarjakin this is the way to remove element from array, so I think I should use `splice` here

Reply:
It doesn't look obvious, can you please add `// NOTE: this.map is being mutated here`